flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/nn/QuadTree.scala

```diff
@@ -0,0 +1,305 @@
+import org.apache.flink.ml.metrics.distances.DistanceMetric
+
+import scala.collection.mutable.ListBuffer
+import scala.collection.mutable.PriorityQueue
+
+/**
+ * n-dimensional QuadTree data structure; partitions
+ * spatial data for faster queries (e.g. KNN query)
+ * The skeleton of the data structure was initially
+ * based off of the 2D Quadtree found here:
+ * http://www.cs.trinity.edu/~mlewis/CSCI1321-F11/Code/src/util/Quadtree.scala
+ *
+ * Many additional methods were added to the class both for
+ * efficient KNN queries and generalizing to n-dim.
+ *
+ * @param minVec
+ * @param maxVec
+ */
+class QuadTree(minVec:ListBuffer[Double], maxVec:ListBuffer[Double],distMetric:DistanceMetric){
+  var maxPerBox = 20
+
+  class Node(c:ListBuffer[Double],L:ListBuffer[Double], var children:ListBuffer[Node]) {
+
+    var objects = new ListBuffer[Vector]
+
+    /** for testing purposes only; used in QuadTreeSuite.scala
+      *
+      * @return
+      */
+    def getCenterLength(): (ListBuffer[Double], ListBuffer[Double]) = {
+      (c, L)
+    }
+
+    def contains(obj: Vector): Boolean = {
+      overlap(obj, 0.0)
+    }
+
+    /** Tests if obj is within a radius of the node
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def overlap(obj: Vector, radius: Double): Boolean = {
+      var count = 0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) - radius < c(i) + L(i) / 2 && obj(i) + radius > c(i) - L(i) / 2) {
+          count += 1
+        }
+      }
+
+      if (count == obj.size) {
+        return true
+      } else {
+        return false
+      }
+    }
+
+    /** Tests if obj is near a node:  minDist is defined so that every point in the box
+      * has distance to obj greater than minDist
+      * (minDist adopted from "Nearest Neighbors Queries" by N. Roussopoulos et al.)
+      *
+      * @param obj
+      * @param radius
+      * @return
+      */
+    def isNear(obj: Vector, radius: Double): Boolean = {
+      if (minDist(obj) < radius) {
+        true
+      } else {
+        false
+      }
+    }
+
+    def minDist(obj: Vector): Double = {
+      var minDist = 0.0
+      for (i <- 0 to obj.size - 1) {
+        if (obj(i) < c(i) - L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) + L(i) / 2, 2)
+        } else if (obj(i) > c(i) + L(i) / 2) {
+          minDist += math.pow(obj(i) - c(i) - L(i) / 2, 2)
+        }
+      }
+      return minDist
+    }
+
+    def whichChild(obj:Vector):Int = {
+
+      var count = 0
+      for (i <- 0 to obj.size - 1){
+        if (obj(i) > c(i)) {
+          count += Math.pow(2,i).toInt
+        }
+      }
+      count
+    }
+
+    def makeChildren() {
+      var cBuff = new ListBuffer[ListBuffer[Double]]
+      cBuff += c
+      var Childrennodes = new ListBuffer[Node]
+      val cPart = partitionBox(cBuff,L,L.length)
+      for (i <- cPart.indices){
+        Childrennodes = Childrennodes :+ new Node(cPart(i), L.map(x => x/2.0), null)
+
+      }
```

Comment:
@tillrohrmann, I had an issue implementing this using Flink's Vector API.  Changing `L` to `width`,  `val mappedWidth = width.map(x => x/2.0)` results in an error.  I searched a bit for this, and a hack I tried `val mappedWidth = width.map(x => x.asInstance[Double]/2.0)`, but that seems to result in a iterable, which cannot be plugged into `Node(center:Vector,width:Vector, var children:ListBuffer[Node])` when constructing `childrenNodes`.